[#121566765] Fix and disable bosh resurrector

keymon opened the pull request:
What
===
We are not sure if we want to use the features covered by Bosh Resurrector. Meanwhile we do [Spike to determine if we want to use it](https://www.pivotaltracker.com/n/projects/1275640/stories/119516757), we will disable the feature.

But the feature was broken because an incorrect password, so we will fix it before disable it.


How to test?
-------

You can check that the new manifest will disable the resurrector because the property `director.hm.resurrector_enabled` is set to `true`:

```
aws s3 cp s3://hector-state/bosh-manifest.yml /tmp
./platform-tests/bosh-template-renderer/render.rb ~/workspace/bosh/release/jobs/health_monitor/templates/health_monitor.yml.erb ~/workspace/bosh/release/jobs/health_monitor/spec /tmp/bosh-manifest.yml bosh | json_pp
``` 

Change the value to see the effect in the template.

First test that resurrector is fixed by:

 1. Create a custom branch reverting the last commit
    ```
git checkout -b resurrector_wip
git revert HEAD
git push
```

 2. Deploy bosh running the pipeline (see README.md)
   ```
BRANCH=$(git rev-parse --abbrev-ref HEAD) make dev pipelines DEPLOY_ENV=...
```

 3. SSH on the bosh machine (via concourse) and check that the logs indicate that resurrector is working:

    ```
$ less /var/vcap/sys/log/health_monitor/health_monitor.log
...
I, [2016-08-11T14:37:07.946532 #19625]  INFO : Adding agent 547e8bff-aa6b-4efc-9b6c-29136e04071b (elasticsearch_master/553475a7-1c63-4633-9ff8-b71bf63bcf46) to hector...
I, [2016-08-11T14:37:07.946579 #19625]  INFO : Adding agent 9cec89fc-1f50-43f7-9546-ffcd86506c90 (elasticsearch_master/38ab835e-9319-452a-95be-47a79930bd37) to hector...
I, [2016-08-11T14:37:07.946629 #19625]  INFO : Adding agent 9b834ea3-8275-4a86-8e29-8da4b36a7945 (maintenance/eff717f7-8ba8-45f6-afda-fc77c3780236) to hector...
I, [2016-08-11T14:37:07.946676 #19625]  INFO : Adding agent e607d566-58ac-425e-92e2-d47dc1f425b2 (rds_broker/71e94328-b348-49c9-a004-440be55ab04b) to hector...
I, [2016-08-11T14:37:07.946721 #19625]  INFO : Adding agent 29b4ab30-f983-4a66-aff9-0bf6e45f98f0 (parser_z2/8259c347-cfce-43b9-be43-4e4ca1e68e6b) to hector...
I, [2016-08-11T14:37:07.946792 #19625]  INFO : Adding agent c26e4b4b-ce6d-476a-a478-7122e89e8c83 (queue/2a1e103a-be3b-49b7-a151-d6950cb1efcf) to hector...
I, [2016-08-11T14:37:07.981550 #19625]  INFO : [ALERT] Alert @ 2016-08-11 14:37:07 UTC, severity 1: process is not running
W, [2016-08-11T14:37:07.981627 #19625]  WARN : (Resurrector) event did not have deployment, job and id: Alert @ 2016-08-11 14:37:07 UTC, severity 1: process is not running
I, [2016-08-11T14:37:08.184654 #19625]  INFO : Managing 1 deployment, 47 agents
I, [2016-08-11T14:37:08.184767 #19625]  INFO : Agent heartbeats received = 170
I, [2016-08-11T14:37:09.033584 #19625]  INFO : [ALERT] Alert @ 2016-08-11 14:37:08 UTC, severity 1: process is not running
W, [2016-08-11T14:37:09.033710 #19625]  WARN : (Resurrector) event did not have deployment, job and id: Alert @ 2016-08-11 14:37:08 UTC, severity 1: process is not running
I, [2016-08-11T14:37:21.900812 #19625]  INFO : Agent `30c2ebbf-e414-4faa-a482-9d14131b61a0' shutting down...
I, [2016-08-11T14:37:21.900906 #19625]  INFO : Removing agent 30c2ebbf-e414-4faa-a482-9d14131b61a0 from all deployments...
```

Second, try to deploy the branch. To do so:

 1. Checkout the original branch `git checkout bugfix/121566765_turn_off_bosh_resurrector`
 2. Redeploy bosh: `BRANCH=$(git rev-parse --abbrev-ref HEAD) make dev pipelines DEPLOY_ENV=...`
 3. check the log, to see if the resurrector is disabled.

Who?
----

Anyone but @keymon

mtekel commented:
Looks like resurrector is not disabled. Maybe the property for global switch is different. It is also hard to tell what's the status of resurrections, there are 2 switches. One global and one per VM. When you do bosh vms/instances it will tell you that resurrection is active on each VM, but that's a per-VM switch. I have not found out how to get global state of resurrection. When you do `bosh status` it gives you state of a lot of things (dns, snapshots etc.), but not resurrection...

dcarley commented:
Not sure it helps, but I remember reading in the [BOSH v257 release notes](https://github.com/cloudfoundry/bosh/releases/tag/stable-3262) (we're using v255.8):

> Keep track of resurrection status globally when it's set via bosh vm resurrection off command
> When temporarily turned off no new resurrection tasks will be enqueued

mtekel commented:
I am not sure what that really means. Perhaps that it would display the real, actual resurrection status for VMs taking the global switch into account? (that is, show that it's disabled when global is disabled?).

Still not sure where to find the status of the global switch...

mtekel commented:
Ok, so you can do `bosh vm resurrection <new global state>`. It accepts `Resurrection paused state should be on/off, true/false, yes/no or enable/disable received "disabled"`. So I set it to off and now all VMs show resurrection state as paused. That's probably the feature described in the release notes, because I remember that before you could have VM resurrection state displayed as active, whereas it was globally disabled.

Nevertheless, I have checked the property in hm [release spec](https://github.com/cloudfoundry/bosh/blob/master/release/jobs/health_monitor/spec) and it doesn't seem like there is typo. Moreover, by default resurrection is disabled. So I am not sure what makes it active in this case...

keymon commented:
I am worried if we are not not disabling it because this code:

https://github.com/cloudfoundry/bosh/blob/master/release/jobs/health_monitor/templates/health_monitor.yml.erb#L121

I am setting the property to "false", but maybe that is string that evaluates `true`. I am trying to test it.

mtekel commented:
Found out I have resurrection: true in my bosh manifest, but already using new HM user. Concourse says it picked the last commit from this branch, so I am going to investigate how is it possible that I have true instead of false...

mtekel commented:
OK, so there is resurrection: true in `bosh-manifest/deployments/aws/090-bosh-aws-stub.yml` - which overrides the false set in this PR...

keymon commented:
I was able to render the health manager config using `bosh-template-renderer` (see description) and I can confirm that setting `resurrector_enabled: false` will disable it.

I removed that from the the stub. 

I also fixed the `bosh-template-renderer` to make allow render the bosh templates.

alext commented:
> I removed that from the the stub.

Sounds like we could do with a unit test for this. It would have caught this issue.

keymon commented:
> Sounds like we could do with a unit test for this. It would have caught this issue.

+1, I will see to implement some testing.

keymon commented:
@alext @mtekel I added the tests to cover the changes in this story.

mtekel commented:
Now it works. There is still a catch that BOSH shows resurrection as `active` when you do `bosh vms --details`, but that's more of a BOSH issue...

merging
